src/pdmfilter.rs

```diff
@@ -0,0 +1,196 @@
+const PDM_DECIMATION: u8 = 64;
+const PI: f32 = 3.14159;
+const SINCN: u8 = 3;
+const FILTER_GAIN: u8 = 64;
+const MAX_VOLUME: u8 = 64;
+
+pub struct PDMFilter {
+    lut: [u32; (SINCN * PDM_DECIMATION / 8) as usize * 256],
+    fs: u32,
+    coef: [u32; 3],
+    sub_const: u32,
+    old_out: i64,
+    old_in: i64,
+    oldz: i64,
+    hp_alpha: u32,
+    lp_alpha: u32,
+    div_const: u32,
+}
+impl PDMFilter {
+    pub fn new(sample_rate: u32) -> PDMFilter {
+        PDMFilter {
+            lut: [0u32; (SINCN * PDM_DECIMATION / 8) as usize * 256],
+            fs: sample_rate * PDM_DECIMATION as u32,
+            coef: [0, 0, 0],
+            sub_const: 0,
+            old_out: 0,
+            old_in: 0,
+            oldz: 0,
+            lp_alpha: 0,
+            hp_alpha: 0,
+            div_const: 0,
+        }
+    }
+    pub fn init(&mut self) {
+        let lp_hz: f32 = self.fs as f32 / 128.0;
+        let hp_hz: f32 = 10.0;
+        if lp_hz != 0.0 {
+            self.lp_alpha = (lp_hz as f32 * 256.0 / (lp_hz + self.fs as f32 / (2.0 * PI))) as u32;
+        }
+        if hp_hz != 0.0 {
+            self.hp_alpha = (self.fs as f32 * 256.0 / (2.0 * PI * hp_hz + self.fs as f32)) as u32;
+        }
```

Comment:
Were there comments in the code that this was ported from/where this came from that could also be copied over to give a high-level overview of the steps involved here?

Reply:
added a comment